dev/docker/ubuntu_20.04_node.dockerfile

```diff
@@ -4,36 +4,59 @@ FROM ubuntu:20.04 as base
 ENV TZ=Canada/Pacific
 RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
 
-# The nodejs that ships on ubuntu:20.04 servers is too old.
-ENV DEBIAN_FRONTEND=noninteractive
-RUN apt-get update -qq
-RUN apt-get install -y --no-install-recommends curl gnupg ca-certificates
-RUN ["/bin/bash", "-c", "set -o pipefail && curl -fsSL https://deb.nodesource.com/setup_lts.x | bash -"]
-RUN apt-get install -y --no-install-recommends nodejs
+# install node with nvm
+RUN apt-get update -qq && apt-get install -y --no-install-recommends git curl wget ca-certificates
+ARG nvm_version="0.39.3"
+ARG node_version="16.20.0"
+RUN mkdir /usr/local/nvm
+ENV NVM_DIR /usr/local/nvm
+ADD https://raw.githubusercontent.com/nvm-sh/nvm/v${nvm_version}/install.sh /nvm_install.sh
+RUN chmod +x /nvm_install.sh && /nvm_install.sh \
+    && . $NVM_DIR/nvm.sh \
+    && nvm install $node_version \
+    && nvm alias default $node_version \
+    && nvm use default
+ENV NODE_PATH $NVM_DIR/v${node_version}/lib/node_modules
+ENV PATH $NVM_DIR/versions/node/v${node_version}/bin:$PATH
+
+# install pnpm
+ENV PNPM_VERSION "7.32.0"
+ADD https://get.pnpm.io/install.sh /tmp/pnpm_install.sh
+RUN chmod +x /tmp/pnpm_install.sh && bash /tmp/pnpm_install.sh
+ENV PNPM_HOME "/root/.local/share/pnpm"
+ENV PATH "${PATH}:${PNPM_HOME}"
 
-# add setup-cpp.js
-COPY "./dist/node12" "/"
-WORKDIR "/"
 
-# run installation
-RUN node ./setup-cpp.js --compiler llvm --cmake true --ninja true --cppcheck true --ccache true --vcpkg true --doxygen true --gcovr true --task true
+#### Building
+FROM base AS builder
+## https://github.com/ever0de/pnpm-docker-root-bug#how-to-fix
+WORKDIR /workspace
+COPY . .
+RUN pnpm install
```

Comment:
I prefer that we built setup-cpp outside Docker. These docker files are meant for C++ development. We can probably use `pnpm install && pnpm pack` outside Docker, and use it inside the image.

I also plan to publish these docker images on DockerHub, so we should start optimizing the docker files (like merging the RUN commands as you mentioned previously).

Reply:
We can use Docker for "integration testing", so maybe we don't need `ubuntu_20.04_node.dockerfile` and `ubuntu_20.04_node_legacy.dockerfile` ... and use docker for (building and) running `setup-cpp.js` + C++-builds.

So we have:
 * arch_node.dockerfile - Building and running on Arch Linux
 * ubuntu_node.dockerfile - Building and running on Ubuntu (+ nvm)
 * fefora_node.dockerfile - Building and running on Fedora
 * ubuntu.dockerfile - Installing setup-cpp (via npm) and runnung